### DISCUSSION TITLE
Help optimizing workflow to reduce burden on CPU

### DISCUSSION THREAD

**Discussion by YUNoSol** [bonsai-rx/bonsai#1677] [Q&A] [answered]
Hi all!

I have created a workflow (with the help of this forum as well) which accomplishes what want.
The issue is that if I'm running the workflow for too long ( around 40 minutes) the pc freezes and crashes with blue screen (stopcode: MACHINE_CHECK_EXCEPTION) 
When I'm recording, the CPU is running at 100%, despite working on a descent PC from what I have gathered, I'm in no way a PC expert. 
Bonsai is the only thing running on the PC when I'm recording.

From what I can find, we could upgrade to a new processor, which might lead to a slight improvement, but no guarantee.
Therefore I was hoping that someone here could spot any redundancies in my workflow, which I could remove, so I can make it more efficient to bring down the burden on the CPU. 

The workflow is displayed here:
![image](https://github.com/bonsai-rx/bonsai/assets/157392652/a024a59e-8353-4fa5-8ac7-aa687dd75a2c)
![image](https://github.com/bonsai-rx/bonsai/assets/157392652/566ca340-e430-45cf-ac1b-896a10007e87)

![image](https://github.com/bonsai-rx/bonsai/assets/157392652/d3fd9cb8-8b89-4264-915a-388402635589)

Attached here:
[20240214.zip](https://github.com/bonsai-rx/bonsai/files/14278870/20240214.zip)

And I need the workflow to be able to do the following:
Record video from 2 different cameras, + audio from a microphone, include timestamps for all three, ensure all three are synchronized, and finally I want to be able to trigger the recording with F5, and for the recording to continue for 10 minutes, and then be ready for another trigger. 
My hope is to be able to run this workflow for at least an hour, preferably 2 hours. 

The specs of the PC I'm running the workflow on currently are the following

NVIDIA RTX A2000
11th gen Intel® Core ™ i7-11700K @ 3.60 GHz
32 GB RAM

I appreciate any help in advance!

Best regards, 
Sol

**Comment by bruno-f-cruz** [comment]
Out of curiosity, what makes you think the problem is with the CPU? A quick look at the pictures you sent makes me think memory accumulation is more likely. Can you confirm this?

**Reply by YUNoSol** [reply]
Hi Bruno

Thank you for your reply

I assumed the CPU was the overburdened part, since it is showing to run at 100% in the task manager as seen below:
![CPU 100%](https://github.com/bonsai-rx/bonsai/assets/157392652/da533d52-dda3-4e67-9e35-bb10ee9da464)

The memory fluctuates between utilizing 18%-63% but has never reached 100%, and the GPU is running at between 18% and 53%, but never 100%, while the CPU is always at 100%
![GPU](https://github.com/bonsai-rx/bonsai/assets/157392652/2a16d474-618a-48ef-bf21-b193a2a75376)

There should be enough memory for storage, 50+ GB last I checked, but I'll of course double check later today.

Best regards, 
Sol

**Reply by bruno-f-cruz** [reply]
I am still not convinced it's not memory tbh. Can you crash it again and keep an eye on the number under 'memory' and report back what you find?

**Reply by YUNoSol** [reply]
Hi Bruno

Of course! I ran the workflow again, and noted down roughly every 5 minutes. I thought you were right at first, since I saw the memory % climb almost linearly as I was recording for 10 minutes and topped at 94 % just before the recording ended. Wierd thing is that I then tried to record for 5 minutes, thinking that that would solve the problem by avoiding the memory climbing too high, but the PC crashed just before ending the recording, with memory at only 53 %. 

My notes are here: 
![image](https://github.com/bonsai-rx/bonsai/assets/157392652/a12a2444-59b6-4e0e-844a-cab6f5c58c2f)

When it crashed yesterday it was after 6 min of a 10 min recording, so it shouldnt be something with the stopping of the recording, and yesterday the workflow ran for about 50 minutes in total before crash, while today it only got to around 30 minutes 

Best regards, 
Sol

**Reply by bruno-f-cruz** [reply]
Can you clarify the following points:
1. Your workflow has several keydown operators. When you state you "ran" the experiment from X amount of time, you are counting from when you click the key?
2. How are you triggering the cameras? Is the frame rate constant? Are there times where cameras go faster than normal (e.g. are they set on auto exposure?)
3. Are both cameras always sending frames?
4. Is there anything else going on the background?
5. Can you remove the stitching of the two frames after `CombineLatest` and try again? Simply take the Item1 from the Tuple outputted from `CombineLatest` instead of using `Concat`.

**Reply by YUNoSol** [reply]
Hi Bruno to your questions:

1. I’m counting the total running time of the workflow from when I press ‘start’ but including time before keydown. 

2. The cameras are hooked up to the PC using USB, and are triggered using the keydown (F5). According to my test using MediaInfo, the framerate should be constant ( 25 fps, no variable framerate detected)

3 I’m unsure about how to confirm this.

4. Only Bonsai, notepad, and sometimes task manager is running at the time of the recording (except one experiment on the 12’th feb. where doric and bonsai were running simultaneously- as seen in the overview below)

5. I’ll give it a try and report back

I have attached a complete overview of my most recent recordings using the workflow, to give you a better picture of when the problem started, and why I have had trouble identifying a pattern.
[6 latest recordings.txt](https://github.com/bonsai-rx/bonsai/files/14310238/6.latest.recordings.txt)

**Reply by YUNoSol** [reply]
Hi again

I tried to remove concat, and use item 1 instead (pictured below) but I got an error after keydown, related to crop later in the workflow.
in an attempt to solve it I removed both crop, but this resulted in both videowriters saving the same video.

![error when running suggested](https://github.com/bonsai-rx/bonsai/assets/157392652/4c1a45aa-75cf-4cd1-8a81-ae957947a46d)

Workflow without crop leading to same video being saved twice
![image](https://github.com/bonsai-rx/bonsai/assets/157392652/03bbff9f-6c62-4329-b986-2ef576e4ef86)


Best regards, 
Sol

**Comment by bruno-f-cruz** [comment]
Can you clarify?


> 1. I’m counting the total running time of the workflow from when I press ‘start’ but including time before keydown.

If the problem is with accumulated frames or some sort of processing bottleneck it would make more sense to count AFTER the frames start pouring in, no?

> 2. The cameras are hooked up to the PC using USB, and are triggered using the keydown (F5). According to my test using MediaInfo, the framerate should be constant ( 25 fps, no variable framerate detected)

This is the encoded frame rate, it will be 25 because that's what you ask from VideoWriter. My question is: How do you know the camera is ACQUIRING at 25fps?

> 
> 3 I’m unsure about how to confirm this.
> 

If they are USB webcames without any sort of GPIO, you either pulse an led in front of them (with known period) or timestamp every single frame (at least it will give you an approximate value of the jitter of the frames)

Your workflow is crashing because your crop is not longer compatible with the frame size.
It makes sense that the same video is saved twice, I just want to know if it still crashes or not.

**Reply by YUNoSol** [reply]
Hi again, sorry for the late answer

> If the problem is with accumulated frames or some sort of processing bottleneck it would make more sense to count AFTER the frames start pouring in, no?

You are right, I'll do that from now on

> This is the encoded frame rate, it will be 25 because that's what you ask from VideoWriter. My question is: How do you know the camera is ACQUIRING at 25fps?

I thought that Mediainfo could analyze the frames, and detect whether or not there was variable or constant framerate, but after analyzing the timestamps, I'm not sure that is the truth.

> If they are USB webcams without any sort of GPIO, you either pulse an led in front of them (with known period) or timestamp every single frame (at least it will give you an approximate value of the jitter of the frames)

I analyzed the timestamps, and found the following, when looking at data from 12 seconds( I assumed this would be enough to get an overview of the frames, I can look at more if necessary)

The number of timestamps varies slightly from second to second. 	
Camera 1	=C1, Camera 2=C2.
	
1’st second: C1 and C2=60 timestamps
2’nd second: C1 and C2=56 TS
3rd – 5’th second: C1 and C2=60 TS
6’th second: 	C1 and C2= 56 TS	
7’th – 10’th second: C1 and C2= 60 TS
11’th second: **C1=57 TS, C2=56 TS**
12’th second: C1 and C2=60 TS

When looking at the milliseconds which each frame was acquired, there was a slight difference between the timestamps of the two cameras.
On average there was a small difference between the two cameras acquisition times (0.059397+-0.081432 milliseconds(ms)) 
Camera 1 sent a TS on average every 0.169075 ms (+-0.036571 ms)
Camera 2 sent a TS on average every 0.169333 ms(+-0.042098 ms).

So I would assume that it would be defined as variable framerate, since it acquires between 56 and 60 timestamps / second

> Your workflow is crashing because your crop is not longer compatible with the frame size.
It makes sense that the same video is saved twice, I just want to know if it still crashes or not.

Using the workflow you suggested (without crop, as imaged above, and linked here)
[20240221-upload.zip](https://github.com/bonsai-rx/bonsai/files/14357772/20240221-upload.zip)

The pc crashed after a total acquisition time of 26 minutes. 
Overview of memory usage in %, when attempting to acquire 3x10 minute recording:
Keydown: 18 %
2 min: 27%
4 min: 35%
6 min: 43%
8 min: 51%
10 min: 59%
-2 min break- 17%
keydown: 17 %
2 min: 25%
4 min: 33%
6 min: 42%
8 min: 50%
10 min: 59%
-few seconds break- 17%
keydown: 17%
2 min: 23%
4 min: 31%
6 min: Crash- did not get to read memory % before crash

I hope my clarification was useful, otherwise let me know, I'm still learning how to best convey this sort of information, so I appreciate your patience.

Best regards, 
Sol

**Reply by YUNoSol** [reply, marked as answer]
Hi again

I'm 99% sure that the problem is the PC, and not the Bonsai workflow.
I'll do an overall health check of the PC, with the help of my IT department. 
Worst case scenario we will use another PC for running Bonsai

Thank you for all the help regardless!

Best regards, 
Sol
